File: app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/adapters/db/dao/CommentDao.java
package it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.Transaction;

import java.util.List;

import it.niedermann.nextcloud.deck.model.ocs.comment.DeckComment;
import it.niedermann.nextcloud.deck.model.ocs.comment.full.FullDeckComment;

@Dao
public interface CommentDao extends GenericDao<DeckComment> {

    @Query("SELECT * FROM DeckComment where accountId = :accountId and id = :remoteId")
    DeckComment getCommentByRemoteIdDirectly(long accountId, Long remoteId);

    @Query("SELECT * FROM DeckComment where accountId = :accountId and localId = :id")
    DeckComment getCommentByLocalIdDirectly(long accountId, Long id);

    @Query("SELECT * FROM DeckComment WHERE accountId = :accountId and objectId = :localCardId and (status<>1 or id is null or lastModified <> lastModifiedLocal)")
    List<DeckComment> getLocallyChangedCommentsByLocalCardIdDirectly(long accountId, long localCardId);

    @Query("SELECT * FROM DeckComment WHERE accountId = :accountId and (status<>1 or id is null or lastModified <> lastModifiedLocal)")
    List<DeckComment> getLocallyChangedCommentsDirectly(long accountId);

    @Query("SELECT * FROM DeckComment WHERE accountId = :accountId and objectId = :localCardId")
    List<DeckComment> getCommentsForLocalCardIdDirectly(long accountId, Long localCardId);

    @Query("SELECT * FROM DeckComment where objectId = :localCardId")
    List<DeckComment> getCommentByLocalCardIdDirectly(Long localCardId);

    @Query("SELECT * FROM DeckComment where objectId = :localCardId order by creationDateTime desc")
    LiveData<List<DeckComment>> getCommentByLocalCardId(Long localCardId);

    @Transaction
    @Query("SELECT * FROM DeckComment where objectId = :localCardId order by creationDateTime desc")
    LiveData<List<FullDeckComment>> getFullCommentByLocalCardId(Long localCardId);

    @Query("SELECT id FROM DeckComment where localId = :localId")
    Long getRemoteCommentIdForLocalIdDirectly(Long localId);

    @Query("SELECT localId FROM DeckComment where id = :remoteId and accountId = :accountId")
    Long getLocalCommentIdForRemoteIdDirectly(long accountId, Long remoteId);
}

File: data/local/src/main/java/it/niedermann/nextcloud/deck/data/local/dao/CommentDao.kt
package it.niedermann.nextcloud.deck.data.local.dao

import androidx.room3.Dao
import androidx.room3.Insert
import androidx.room3.OnConflictStrategy
import androidx.room3.Query
import io.reactivex.rxjava3.core.Flowable
import it.niedermann.nextcloud.deck.data.local.entity.CommentEntity
import java.util.concurrent.CompletableFuture

@Dao
interface CommentDao : GenericDao<CommentEntity> {

    @Query("SELECT * FROM Comment WHERE cardId = :cardId")
    fun getCommentsByCard(cardId: Long): Flowable<List<CommentEntity>>

    @Query("SELECT * FROM Comment WHERE cardId = :cardId")
    fun getCommentsByCardRx(cardId: Long): CompletableFuture<List<CommentEntity>>

    @Query("SELECT * FROM Comment WHERE accountId = :accountId AND status != 1")
    fun getChangedComments(accountId: Long): CompletableFuture<List<CommentEntity>>

    @Query("SELECT * FROM Comment WHERE accountId = :accountId AND remoteId = :remoteId")
    fun getCommentByRemoteId(accountId: Long, remoteId: Long): CompletableFuture<CommentEntity?>

    @Query("SELECT * FROM Comment WHERE localId = :localId")
    fun getCommentByLocalId(localId: Long): CompletableFuture<CommentEntity?>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertOrReplace(entity: CommentEntity): CompletableFuture<Long>

    @Query("DELETE FROM Comment WHERE localId = :localId")
    fun deleteById(localId: Long): CompletableFuture<Void?>
}
